cardano-db-sync/src/Cardano/DbSync/Cache/Epoch.hs

```diff
@@ -0,0 +1,90 @@
+module Cardano.DbSync.Cache.Epoch (
+  readCacheEpoch,
+  readEpochFromCacheEpoch,
+  writeCacheEpoch,
+  writeBlockAndFeeToCacheEpoch,
+  writeEpochToCacheEpoch,
+) where
+
+import qualified Cardano.Db as DB
+import Cardano.DbSync.Cache.Types (Cache (..), CacheEpoch (..), CacheInternal (..))
+import Cardano.DbSync.Types (CardanoBlock, EpochSlot (..))
+import Cardano.Prelude
+import Cardano.Slotting.Slot (EpochNo (..))
+import Control.Concurrent.Class.MonadSTM.Strict (readTVarIO, writeTVar)
+import Database.Persist.Postgresql (SqlBackend)
+
+-------------------------------------------------------------------------------------
+-- Epoch Cache
+-------------------------------------------------------------------------------------
+readCacheEpoch :: Cache -> IO (Maybe CacheEpoch)
+readCacheEpoch cache =
+  case cache of
+    UninitiatedCache -> pure Nothing
+    Cache ci -> readTVarIO (cEpoch ci)
+
+readEpochFromCacheEpoch :: Cache -> IO (Maybe DB.Epoch)
+readEpochFromCacheEpoch cache =
+  case cache of
+    UninitiatedCache -> pure Nothing
+    Cache ci -> do
+      cachedEpoch <- readTVarIO (cEpoch ci)
+      case cachedEpoch of
+        Nothing -> pure Nothing
+        Just ce -> pure $ Just =<< ceEpoch ce
+
+writeCacheEpoch ::
+  MonadIO m =>
+  Cache ->
+  CacheEpoch ->
+  ReaderT SqlBackend m ()
+writeCacheEpoch cache cacheEpoch =
+  case cache of
+    UninitiatedCache -> pure ()
+    Cache ci -> liftIO $ atomically $ writeTVar (cEpoch ci) $ Just cacheEpoch
+
+writeBlockAndFeeToCacheEpoch ::
+  MonadIO m =>
+  Cache ->
+  CardanoBlock ->
```

Comment:
Like in the comments above, instead of `CardanoBlock` you could pass directly the fields that you need. Or alternatively the `Generic.Block`. This can avoid us duplicating the logic.